### ISSUE TITLE
Provide a way to detect infinite component rendering recursion in development

### ISSUE THREAD

**Issue by josh-degraw** [facebook/react#12525]
**Do you want to request a *feature* or report a *bug*?**
- Feature (possibly bug?)

**What is the current behavior?**

I've been trying out the new Context API in my project and it's awesome. However, in my haste to start using it, I managed to stumble into a situation where every time I would try and render a certain component which was making use of a few different contexts, the app would completely freeze, and the only thing that would let me get out of this error state was to forcefully kill the process via the chrome task manager.

Nothing would be logged to the console, the app would just completely freeze, and when I opened up the task manager and saw the CPU spiked up every time i would go to this component, and the only way I could stop it was to crash the tab. 

I finally threw some `console` statements in and saw that it had just entered into an infinite loop between these providers. I managed to get the app to stop crashing, but I'm still unsure as to why exactly this was happening. I'm sure I was just using this API incorrectly somehow, but this was a very confusing problem to diagnose, and some error checking here would be incredibly useful

**What is the expected behavior?**

It would be very beneficial to have some sort of checks in place, similar to what happens with too many `setState` calls happening too closely when you call it from `componentDidUpdate`, for example. That way, instead of freezing everything up permanently, the app could at least crash and report some sort of information and help me realize where I'd gone wrong.

**Which versions of React, and which browser / OS are affected by this issue? Did this work in previous versions of React?**
- React `16.3.0`
- Chrome `65.0.3325.181`

**Comment by gaearon** [comment]
Is there something specific to context here? I guess you'd have a similar problem if you rendered a component that recursively renders itself?

**Comment by awnigharbia** [comment]
@gaearon  Is the same issue I post it before two days? here?

**Comment by josh-degraw** [comment]
Yeah that's basically what was going on, it was just a recursive call with no end condition, and I just didn't realize as I was writing it. 

It's probably more that I was using HOC functions to do this, and those functions were recursively calling each other without me realizing.

I had set up a few different functions like this to combine common contexts:

```jsx

const AppContextTypeContext = React.createContext('None');
const AppContextIdContext = React.createContext(0);

function AppContextProvider({ contextType, contextId, ...props }) {
    return (
        <AppContextTypeContext.Provider value={contextType}>
            <AppContextIdContext.Provider value={contextId} {...props} />
        </AppContextTypeContext.Provider>
    );  
}
```
with matching HOC functions for the context like this:
```jsx
const withAppContext = WrappedComponent => function WithAppContext(props) {
    return (
        <AppContextTypeContext.Consumer>
            {contextType => (
                <AppContextIdContext.Consumer>
                    {contextId => <WrappedComponent contextType={contextType} contextId={contextId} {...props} />}
                </AppContextIdContext.Consumer>
            )}
        </AppContextTypeContext.Consumer>);
};
```
and I was combining them in multiple ways in multiple spots, and I guess I just ended up with the wrong combination of these functions that led to this issue.

So in short yeah, it's probably not specific just to contexts, but I could see many others easily running into this issue like I did.

**Comment by nickmccurdy** [comment]
This does seem to be a duplicate of #12515, though it's been a helpful discussion so far.

**Comment by josh-degraw** [comment]
Honestly I think it really is effictively just a more oundabout format of your second example. It's easy to see when it's like that, but it gets more difficult to diagnose when you have used 3 or 4 HOCs on top of otherwise unrelated components that then happen to call each other. 

I guess what I'm requesting here is really just some sort of measure to detect this and throw an error, and break out of that loop somehow, so recursive issues such as this can be diagnosed easier, much like how you handle too many `setState` calls occurring too close together.

**Comment by nickmccurdy** [comment]
I think the issue is technically unrelated to context, though it seems to be easier to run into with the new context API. Is there any built in safeguard to prevent too much recursion or would a check for this be easy to add at least for debugging purposes?

**Comment by nmain** [comment]
What would be the desired solution here; keep track of how deep components have nested during rendering and abort with message when it's too deep?  (200?  500?)

**Comment by nickmccurdy** [comment]
I was originally thinking about warning when a component's child renders another instance of itself, but that could have confusing false positives. I guess setting a max depth is a good compromise, if it only warns in dev.

**Comment by antwan2093** [comment]
It’s not necessarily what you’re looking for, but I find the npm package “why did you update” to be useful in identifying unnecessary rerenders.

**Comment by giraffesyo** [comment]
This is basically the [halting problem](https://en.wikipedia.org/wiki/Halting_problem), a program cannot be created that will determine if another program will ever stop running. However, you can set some form of counter and after so many iterations halt a program.

I don't think it is a something that should be added as it's the same idea across all programming languages, if you keep calling a function infinitely, your program will run infinitely.

**Comment by sahilrajput03** [comment]
```jsx
//Custom hook
const useRenderCount = () => {
  // src: https://medium.com/better-programming/how-to-properly-use-the-react-useref-hook-in-concurrent-mode-38c54543857b
  const renderCount = useRef(0);
  let renderCountLocal = renderCount.current;
  useEffect(() => {
    renderCount.current = renderCountLocal;
  });
  renderCountLocal++;
  return renderCount.current;
};
```
And using that custom hook in the function component like, 
```jsx
  const App () => {
  const renderCount = useRenderCount();

  if (renderCount > 10) throw new Error();//This works great for me.

  //...more code..
  }
```

**Comment by eps1lon** [comment]
Re-opening since it came up again in https://github.com/facebook/react/issues/20629

**Comment by gla23** [comment]
I struggled to debug my infinite loop because the chrome tab would crash (which took ages to close and reload) and the component tree wouldn't render (to display why the loop was happening). A helpful tip to get round these two issues is to use the following:
```js
if (Math.random() < 0.2) return "infinite loop ended here";
```

**Comment by NoamAnisfeld** [comment]
Just bumped into this, too - the cause of the issue is ridiculously simple but it's very hard to debug due to the unclear freeze.

Sure as mentioned, theoretically there's no way to detect an infinite loop. But in practice you could set it to stop and output an error when the recursion depth reaches some very high value - just like vanilla JS does. Or if stopping is not acceptable, at least output a warning when in development and strict mode.
